Distribute iminuit as a wheel on PyPI

It would be quite nice if `iminuit` would distribute wheels for the library on PyPI. [Wheels](https://wheel.readthedocs.io/en/latest/) make installation much faster and given that not everyone wants to use a very bulky system like Conda (which is [currently suffering terrible speed issues](https://www.anaconda.com/understanding-and-improving-condas-performance/) that [mamba](https://github.com/QuantStack/mamba) may or may not fix) it is a great resource for many users to have binary dists available.

At the moment it appears that there are issues with the way that iminuit's CI is setup that the building of wheels is failing. For example, see the following from a [Python 3.6 job on Travis CI](https://travis-ci.org/scikit-hep/iminuit/jobs/528537856#L736-L784)

```
  x86_64-linux-gnu-gcc -pthread -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fPIC -DWARNINGMSG=1 -I/tmp/pip-req-build-IViZEC/Minuit/inc -I/usr/include/python2.7 -c /tmp/pip-req-build-IViZEC/iminuit/_libiminuit.cpp -o build/temp.linux-x86_64-2.7/tmp/pip-req-build-IViZEC/iminuit/_libiminuit.o
  cc1plus: warning: command line option ‘-Wstrict-prototypes’ is valid for C/ObjC but not for C++ [enabled by default]
  In file included from /tmp/pip-req-build-IViZEC/iminuit/PythonFCN.h:9:0,
                   from /tmp/pip-req-build-IViZEC/iminuit/_libiminuit.cpp:723:
  /tmp/pip-req-build-IViZEC/iminuit/PythonCaller.h:10:31: fatal error: numpy/arrayobject.h: No such file or directory
   #include "numpy/arrayobject.h"
                                 ^
  compilation terminated.
  error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
  ----------------------------------------
  ERROR: Failed building wheel for iminuit
```

However, if these issues can be fixed then the good news is that having the CI build and distribute wheels is [incredibly simple](https://github.com/diana-hep/pyhf/blob/b6edffae1a1cb7fc7544052506ad790a81960171/.travis.yml#L145).

If this would be considered in the future I would be willing to give time, as available, to help out on this.

I've got this working on:

| System | Arch | Python versions |
|---------|-----|------------------|
| ManyLinux1 (custom GCC 8.3) | 64 & 32-bit | 2.7, 3.5, 3.6, 3.7 |
| ManyLinux2010 | 64-bit | 2.7, 3.5, 3.6, 3.7 |
| macOS 10.9+ | 64-bit | 2.7, 3.6, 3.7 |
| Windows | 64 & 32-bit | 2.7, 3.6, 3.7 |

This comes from boost-histogram; if there's a reason to have it, 3.5 on Windows should be eas(ier) to add (than it would have been for boost-histogram).

Only issue is 32-bit Windows + Python 3.6 or 3.7, where one test fails (I *think* the wheels are fine):

```
================================== FAILURES ===================================
________________________________ test_ackleys _________________________________

    def test_ackleys():
        random.seed(0.258)
    
        m = Minuit(ackleys, x=1.5 * random.random(), y=1.5 * random.random(),
                   error_x=1.7, error_y=1.7,
                   pedantic=False, print_level=0)
        m.tol = 1e-4
        m.migrad()
    
>       assert_allclose(m.fval, 0, atol=1e-6)
E       AssertionError: 
E       Not equal to tolerance rtol=1e-07, atol=1e-06
E       
E       Mismatch: 100%
E       Max absolute difference: 2.57992756
E       Max relative difference: inf
E        x: array(2.579928)
E        y: array(0)

C:\hostedtoolcache\windows\Python\3.6.8\x86\lib\site-packages\iminuit\tests\test_functions.py:39: AssertionError
============================== warnings summary ===============================
tests/test_minimize.py::test_simple
tests/test_minimize.py::test_simple
tests/test_minimize.py::test_simple
  C:\hostedtoolcache\windows\Python\3.6.8\x86\lib\importlib\_bootstrap.py:219: RuntimeWarning: numpy.ufunc size changed, may indicate binary incompatibility. Expected 112 from C header, got 124 from PyObject
    return f(*args, **kwds)
```

--- a/.ci/azure-build-pipeline.yml
+++ b/.ci/azure-build-pipeline.yml
@@ -1,0 +1,104 @@
+trigger:
+  tags:
+    include:
+    - v*
+
+pr:
+- master
+
+
+# Required variables (set in UI):
+#   package_name:     This is the output name, - is replaced by _
+#   many_linux_base:  Should be quay.io/pypa/manylinux1_ or skhep/manylinuxgcc-
+variables:
+  package_name: iminuit
+  many_linux_base: quay.io/pypa/manylinux1_
+
+jobs:
+
+- job: LinuxSDist
+  pool:
+    vmImage: 'ubuntu-16.04'
+  variables:
+    python.architecture: 'none'
+  steps:
+    - script: |
+        python -m pip install setuptools cython
+        python setup.py sdist
+      displayName: Publish sdist
+    - template: azure-publish-dist.yml
+
+- job: ManyLinux
+  strategy:
+    matrix:
+      64Bit2010:
+        arch: x86_64
+        plat: manylinux2010_x86_64
+        image: quay.io/pypa/manylinux2010_x86_64
+        python.architecture: x64
+      64Bit:
+        arch: x86_64
+        plat: manylinux1_x86_64
+        image: $(many_linux_base)x86_64
+        python.architecture: x64
+      32Bit:
+        arch: i686
+        plat: manylinux1_i686
+        image: $(many_linux_base)i686
+        python.architecture: x86
+  pool:
+    vmImage: 'ubuntu-16.04'
+  steps:
+    - script: |
+        docker run -e NPY_NUM_BUILD_JOBS=4 -e PLAT=$(plat) -e package_name=$(package_name) --rm -v `pwd`:/io $(image) /io/.ci/build-wheels.sh
+        ls -lh wheelhouse/
+        mkdir -p dist
+        cp wheelhouse/$(package_name)*.whl dist/.
+      displayName: Build wheels
+    - template: azure-publish-dist.yml
+
+- job: macOS
+  variables:
+    python.architecture: 'x64'
+  strategy:
+    matrix:
+      Python27:
+        python.version: '2.7'
+      Python36:
+        python.version: '3.6'
+      Python37:
+        python.version: '3.7'
+  pool:
+    vmImage: 'macOS-10.14'
+  steps:
+    - template: azure-setup.yml
+    - template: azure-steps.yml
+    - template: azure-publish-dist.yml
+
+- job: Windows
+  strategy:
+    matrix:
+      Python27:
+        python.version: '2.7'
+        python.architecture: 'x64'
+      Python36:
+        python.version: '3.6'
+        python.architecture: 'x64'
+      Python37:
+        python.version: '3.7'
+        python.architecture: 'x64'
+      Python27_32:
+        python.version: '2.7'
+        python.architecture: 'x86'
+      Python36_32:
+        python.version: '3.6'
+        python.architecture: 'x86'
+      Python37_32:
+        python.version: '3.7'
+        python.architecture: 'x86'
+  pool:
+    vmImage: 'vs2017-win2016'
+  steps:
+    - template: azure-setup.yml
+    - template: azure-steps.yml
+    - template: azure-publish-dist.yml

--- a/.ci/azure-publish-dist.yml
+++ b/.ci/azure-publish-dist.yml
@@ -1,0 +1,9 @@
+
+steps:
+- task: PublishPipelineArtifact@0
+  inputs:
+    artifactName: '$(Agent.OS)_$(Agent.JobName)_$(python.architecture)'
+    targetPath: 'dist'
+
+# You can set FULLBUILD_FORCE to anything non-blank to get the manylinux job to run.
+# Artifacts are attached to the build, and are published with an Azure release pipeline.

--- a/.ci/azure-setup.yml
+++ b/.ci/azure-setup.yml
@@ -1,0 +1,22 @@
+# This is special for macOS because "UsePythonVersion"
+# does not use the official Python.org installers.
+
+steps:
+
+- script: .ci/macos-install-python.sh '$(python.version)'
+  displayName: Install Python.org Python
+  condition: and(succeeded(), eq(variables['Agent.OS'], 'Darwin')) 
+
+- task: UsePythonVersion@0
+  inputs:
+    versionSpec: '$(python.version)'
+    architecture: '$(python.architecture)'
+  condition: and(succeeded(), ne(variables['Agent.OS'], 'Darwin')) 
+
+- script: |
+    mkdir -p dist
+    python -m pip install --upgrade pip
+    python -m pip install --upgrade pytest wheel twine setuptools
+    python -m pip install -r dev-requirements.txt
+  displayName: 'Install dependencies'
+

--- a/.ci/azure-steps.yml
+++ b/.ci/azure-steps.yml
@@ -1,0 +1,42 @@
+steps:
+
+- script: |
+    call "C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Auxiliary\Build\vcvarsall.bat" $(python.architecture)
+    set MSSdk=1
+    set DISTUTILS_USE_SDK=1
+    python -m pip wheel . -w wheelhouse/
+  displayName: 'Build wheel (Windows Python 2.7)'
+  condition: and(succeeded(), eq(variables['Agent.OS'], 'Windows_NT'), eq(variables['python.version'], '2.7')) 
+
+- script: |
+    python -m pip wheel . -w wheelhouse/
+  displayName: 'Build wheel'
+  condition: and(succeeded(), not(and(eq(variables['Agent.OS'], 'Windows_NT'), eq(variables['python.version'], '2.7'))))
+
+- script: |
+    ls -lh wheelhouse
+    mkdir -p dist
+    cp wheelhouse/$(package_name)* dist/.
+  displayName: 'Show wheelhouse'
+
+- script: |
+    python -m pip install $(package_name) --no-index -f wheelhouse
+  displayName: 'Install wheel'
+
+- script: |
+    python -m pytest --pyargs iminuit --junitxml=junit/test-results.xml
+  # I'm selecting a random directory to not trigger relative imports.
+  workingDirectory: tutorial
+  displayName: 'Test with pytest'
+
+- task: PublishTestResults@2
+  inputs:
+    testResultsFiles: '**/test-*.xml'
+    testRunTitle: 'Publish test results for Python $(python.version)'
+  condition: succeededOrFailed()
+
+- script: |
+    python -m pip install delocate
+    /Library/Frameworks/Python.framework/Versions/$(python.version)/bin/delocate-wheel dist/$(package_name)*.whl
+  displayName: 'Delocate wheels'
+  condition: and(succeeded(), eq(variables['Agent.OS'], 'Darwin')) 

--- a/.ci/build-wheels.sh
+++ b/.ci/build-wheels.sh
@@ -1,0 +1,25 @@
+#!/bin/bash
+set -e -x
+
+# Collect the pythons
+pys=(/opt/python/*/bin)
+
+# Filter out Python 3.4
+pys=(${pys[@]//*34*/})
+
+# Compile wheels
+for PYBIN in "${pys[@]}"; do
+    "${PYBIN}/pip" install -r /io/dev-requirements.txt
+    "${PYBIN}/pip" wheel /io/ -w wheelhouse/
+done
+
+# Bundle external shared libraries into the wheels
+for whl in wheelhouse/$package_name-*.whl; do
+    auditwheel repair --plat $PLAT "$whl" -w /io/wheelhouse/
+done
+
+# Install packages and test
+for PYBIN in "${pys[@]}"; do
+    "${PYBIN}/python" -m pip install $package_name --no-index -f /io/wheelhouse
+    "${PYBIN}/pytest" /io/tests
+done

--- a/.ci/macos-install-python.sh
+++ b/.ci/macos-install-python.sh
@@ -1,0 +1,34 @@
+#!/usr/bin/env bash
+
+PYTHON_VERSION="$1"
+
+case $PYTHON_VERSION in
+2.7)
+  FULL_VERSION=2.7.16
+  ;;
+3.6)
+  FULL_VERSION=3.6.8
+  ;;
+3.7)
+  FULL_VERSION=3.7.3
+  ;;
+esac
+
+INSTALLER_NAME=python-$FULL_VERSION-macosx10.9.pkg
+URL=https://www.python.org/ftp/python/$FULL_VERSION/$INSTALLER_NAME
+
+PY_PREFIX=/Library/Frameworks/Python.framework/Versions
+
+set -e -x
+
+curl $URL > $INSTALLER_NAME
+
+sudo installer -pkg $INSTALLER_NAME -target /
+
+sudo rm /usr/local/bin/python
+sudo ln -s /usr/local/bin/python$PYTHON_VERSION /usr/local/bin/python
+
+which python
+python --version
+python -m ensurepip
+python -m pip install setuptools twine wheel numpy

--- a/dev-requirements.txt
+++ b/dev-requirements.txt
@@ -1,0 +1,10 @@
+numpy
+scipy
+
+cython
+
+pytest
+pytest-cov
+pylint
+flake8
+pydocstyle

